test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Kickstart_default_finish.host4dhcp.snap.txt

```diff
@@ -78,12 +78,11 @@ echo "Performing initial puppet run for --tags no_such_tag"
 
 
 
-cat << EOF | sed 's/^ //' > /root/ansible_provisioning_call.sh
- #!/bin/sh
- 
- echo "Calling Ansible AWX/Tower provisioning callback..."
- /usr/bin/curl -v -k -s --data "host_config_key=" https:///api/v2/job_templates//callback/
- echo "DONE"
+cat << EOF | base64 -d > /root/ansible_provisioning_call.sh
+IyEvYmluL3NoCgplY2hvICJDYWxsaW5nIEFuc2libGUgQVdYL1Rvd2VyIHBy
```

Comment:
It does make it harder to review what's really going on.

Reply:
That's true. I wouldn't mind doing it differently. Any suggestions?

Reply:
I'm looking at the original issue and wonder what's going on. One possible solution is to use a different string than `EOF` (given it's quite common). For example `FOREMAN_WRITE_CONTENT`. It reduces the risk.

Another thing to consider is that you may actually change functionality. Because it's base64 that means bash won't replace variables. It can be that something relies on that functionality.

Other solutions that come to mind are making base64 behavior opt-in. So `save_to_file(filename, content, base64: false)`.

A third option that is specific to REX setup is skip the whole write-to-file-and-then-execute-that-file approach. Just run it literally. You are already in a shell, right? Or is it intentional that you want to leave the script on disk?

Reply:
> For example FOREMAN_WRITE_CONTENT. It reduces the risk

Reduces, but does not remove it completely. How about having it a bit dynamic and adding a, possibly random, suffix to it unless we arrive at a delimiter that is not present in the thing being saved?

> Because it's base64 that means bash won't replace variables. It can be that something relies on that functionality.

To me the variable replacing behaviour was surprising, but sadly, it's possible that something somewhere depends on it. How about a separate helper or a flag to signify if the content should be placed to the disk literally?

>  You are already in a shell, right?

Well, yes, but the same template is used in different contexts (rex job, provisioning templates), in some of them running it in a separate shell might be desirable (the rex pull setup snippet may call exit in some cases)

Reply:
> How about having it a bit dynamic and adding a, possibly random, suffix to it unless we arrive at a delimiter that is not present in the thing being saved?

Random will mess with the snapshots. You could hash the filename and/or content though and add that to it.

> To me the variable replacing behaviour was surprising, but sadly, it's possible that something somewhere depends on it. How about a separate helper or a flag to signify if the content should be placed to the disk literally?

That's kind of what my second suggestion would do. Whether you name it `base64` or something else can be considered an implementation detail.

> Well, yes, but the same template is used in different contexts (rex job, provisioning templates), in some of them running it in a separate shell might be desirable (the rex pull setup snippet may call exit in some cases)

That's a fair point.